File: resources/features/steps/apportionment_steps.py
from behave import Given, When, Then
from python.sml.methods.apportionment import apportionment


# Given("""^there are datasets at the locations:$""")

    # Do common step


# Given("""^the user provides the parameters:$""")

    # Do common step


@Given("there is a simple relationship between two partitions")
def given_simple_relationship(context):

    # Passing step, functionality tested by feature examples
    pass


@Given("there is a complex relationship between two partitions")
def given_complex_relationship(context):

    # Passing step, functionality tested by feature examples
    pass


@When("the Python Apportionment function is applied to the dataset")
def when_apportioning(context):

    # Load parameters, create DataFrame and apply Apportionment function with parameters
    context.params = context.data_row + context.param_row

    input_df = context.spark                    \
                      .read                     \
                      .json(context.params[0])

    context.result_df = apportionment(input_df).app1(context.params[1],
                                                     context.params[2],
                                                     context.params[3],
                                                     context.params[4],
                                                     context.params[5].split(","))


@Then("the Apportionment function will apportion the value to the output, matching an expected dataset at a location {expected_path}")
def then_assert_df_equivalence(context, expected_path):

    # Assert that the output DataFrame matches a DataFrame created from the expected data
    expected_df = context.spark                 \
                         .read                  \
                         .json(expected_path)

    # Pull out columns from expected DataFrame to re-order columns in output DataFrame
    select_list = expected_df.columns

    assert context.result_df                        \
                  .select(select_list)              \
                  .orderBy("vatref9", "turnover")   \
                  .collect()                        \
          ==                                        \
          expected_df.select(select_list)           \
                     .orderBy("vatref9", "turnover")\
                     .collect()

@Then("the Apportionment function ensures that no turnover is lost in the output")
def then_assert_no_turnover_loss(context):

    # Assert that turnover is not lost for each aggregate B reference
    target0 = context.params[5].split(",")[0]
    agg_b = context.params[4]

    for reference in context.result_df.select(agg_b).distinct().collect()[0]:
        assert context.result_df                        \
                      .filter(agg_b + "=" + reference)  \
                      .groupBy(agg_b)                   \
                      .sum(target0)                     \
                      .select("sum(" + target0 + ")")   \
                      .first()                          \
               ==                                       \
               context.result_df                        \
                      .filter(agg_b + "=" + reference)  \
                      .select(target0 + "_apportioned") \
                      .first()

File: src/test/scala/uk/gov/ons/stepdefs/ApportionmentSteps.scala
package uk.gov.ons.stepdefs

import cucumber.api.scala.{EN, ScalaDsl}
import org.apache.spark.sql.DataFrame
import uk.gov.ons.methods.Apportionment
import scala.collection.JavaConverters._

class ApportionmentSteps extends ScalaDsl with EN {

  /**
    * Combines the first element (Java util.Map[String, String]) of two Java util.Lists and converts to a
    * single Scala Map[String, String].
    *
    * @param list1 : util.List[util.Map[String, String]\]
    * @param list2 : util.List[util.Map[String, String]\]
    */
  def convertJavaListsOfMaps(list1: java.util.List[java.util.Map[String, String]],
                             list2: java.util.List[java.util.Map[String, String]]): Map[String, String] = {
    List.concat(list1.get(0).asScala, list2.get(0).asScala).toMap
  }

  // Given("""^there are datasets at the locations:$""")
    // Do common step

  // Given("""^the user provides the parameters:$""")
    // Do common step

  Given("""^there is a simple relationship between two partitions$""") { () =>
    // Passing step, functionality tested by feature examples
    println("Given there is a simple relationship between two partitions")
  }

  Given("""^there is a complex relationship between two partitions$""") { () =>
    // Passing step, functionality tested by feature examples
    println("Given there is a complex relationship between two partitions")
  }

  When("""^the Scala Apportionment function is applied to the dataset$""") { () =>
    // Load parameters, create DataFrame and apply Apportionment function with parameters
    ContextApportionment.paramsMap = convertJavaListsOfMaps(ContextCommon.paths, ContextCommon.params)
    val inputDF = Helpers.sparkSession
                         .read
                         .json(ContextApportionment.paramsMap("input_data"))
    ContextApportionment.resultDF = Apportionment.apportionment(inputDF)
                                                 .app1(ContextApportionment.paramsMap("aux"),
                                                       ContextApportionment.paramsMap("date"),
                                                       ContextApportionment.paramsMap("aggregate A"),
                                                       ContextApportionment.paramsMap("aggregate B"),
                                                       ContextApportionment.paramsMap("apportion").split(",").toList)
    println("When the Scala Apportionment function is applied to the dataset")
  }

  Then(
    """^the Apportionment function will apportion the value to the output, matching an expected dataset at a location ([^"]*)$""".stripMargin) { (arg0: String) =>
    // Assert that the output DataFrame matches a DataFrame created from the expected data
    val expectedDF = Helpers.sparkSession
                            .read
                            .json(arg0)
    // Pull out columns from expected DataFrame to re-order columns in output DataFrame
    val selectList = expectedDF.columns

    assert(ContextApportionment.resultDF
                               .select(selectList.head, selectList.tail: _*)
                               .orderBy("vatref9", "turnover")
                               .collect()
      sameElements
           expectedDF.select(selectList.head, selectList.tail: _*)
                     .orderBy("vatref9", "turnover")
                     .collect())
    println("Then the Apportionment function will apportion the value to the output, matching an expected dataset " + "at a location " + arg0)
  }

  Then(
    """^the Apportionment function ensures that no turnover is lost in the output$""") { () =>
    // Assert that turnover is not lost for each aggregate B reference
    val target0 = ContextApportionment.paramsMap("apportion").split(",").toList(0)
    val aggB = ContextApportionment.paramsMap("aggregate B")

    for (refs <- ContextApportionment.resultDF.select(aggB).distinct().collect()) {
      val A = ContextApportionment.resultDF
                                  .filter(aggB + "=" + refs(0))
                                  .groupBy(aggB)
                                  .sum(target0)
                                  .select("sum(" + target0 + ")")
                                  .first()
      val B = ContextApportionment.resultDF
                                  .filter(aggB + "=" + refs(0))
                                  .select(target0 + "_apportioned")
                                  .first()
      assert(A == B)
    }
    println("Then the Apportionment function ensures that no turnover is lost in the output")
  }
}

object ContextApportionment {
  var resultDF: DataFrame = _
  var paramsMap: Map[String, String] = _
}
